src/test_double_builder.ts

```diff
@@ -0,0 +1,174 @@
+import type { Constructor } from "./types.ts";
+
+/**
+ * This class contains the methods to help build most of what a test double
+ * needs.
+ */
+export class TestDoubleBuilder<OriginalClass> {
+  /**
+   * The class object passed into the constructor
+   */
+  protected constructor_fn: Constructor<OriginalClass>;
+
+  /**
+   * A list of arguments the class constructor takes
+   */
+  protected constructor_args: unknown[] = [];
+
+  protected native_methods = [
+    "__defineGetter__",
+    "__defineSetter__",
+    "__lookupGetter__",
+    "__lookupSetter__",
+    "constructor",
+    "hasOwnProperty",
+    "isPrototypeOf",
+    "propertyIsEnumerable",
+    "toLocaleString",
+    "toString",
+    "valueOf",
+  ];
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - CONSTRUCTOR ///////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Construct an object of this class.
+   *
+   * @param constructorFn - The constructor function of the object to create a
+   * test double out of.
+   */
+  constructor(constructorFn: Constructor<OriginalClass>) {
+    this.constructor_fn = constructorFn;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - PROTECTED /////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Get all functions from the original so they can be added to the test
+   * double.
+   *
+   * @param obj - The object that will be a test double.
+   *
+   * @returns An array of the object's functions.
+   */
+  protected getAllFunctionNames(obj: OriginalClass): string[] {
+    let functions: string[] = [];
+    let clone = obj;
+    do {
+      functions = functions.concat(Object.getOwnPropertyNames(clone));
+    } while ((clone = Object.getPrototypeOf(clone)));
```

Comment:
im guessing this works but i have no clue how, surely this would be an infinite loop?

Reply:
so the clone can be an extended class (can be an extended class of many extended classes). this loop gets all of the functions from each prototype so that the mock can have all the inherited functions present. here's an example script:

```ts
class A {}
class B extends A {}
class C extends B {}
class D extends C {}

let clone = new D();

do {
  console.log(clone);
  // Iteration 1 ---> D {}
  // Iteration 2 ---> D {}
  // Iteration 3 ---> C {}
  // Iteration 4 ---> B {}
  // Iteration 5 ---> A {}
  // Iteration 6 ---> {}   // Stops here
} while ((clone = Object.getPrototypeOf(clone)));
```

Reply:
i should add a comment abotu this lol

Reply:
thanks, im just an idiot lol

Reply:
haha nah. comment added tho! [34cd6ff](https://github.com/drashland/rhum/pull/163/commits/34cd6ffedfe622e3ab787881f3ba207cb90c9b9d)

Reply:
niceeeee